[PERF] survey: improve page switching performance

flch-odoo opened the pull request:
Validating answers when completing a survey is sometimes slow.

This PR reduces the number of queries required to validate answers 
and see the next question by 6.5% to 30% depending on the endpoint, 
question type and number of answers selected in multiple choice 
questions.

Task-3742599

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/153721).

std-odoo commented on addons/survey/controllers/main.py:
```diff
@@ -29,15 +29,24 @@ def _fetch_from_access_token(self, survey_token, answer_token):
         Returns a sudo-ed browse record of survey in order to avoid access rights
         issues now that access is granted through token. """
-        survey_sudo = request.env['survey.survey'].with_context(active_test=False).sudo().search([('access_token', '=', survey_token)])
+        if not survey_token:
```
why not something like that ? (it has the same query count, and we don't bypass the ORM)

```py
def _fetch_from_access_token(self, survey_token, answer_token):
    if not survey_token:
        return request.env['survey.survey'].sudo(), request.env['survey.user_input'].sudo()

    if not answer_token:
        s = request.env['survey.survey'].sudo().search([('access_token', '=', survey_token)], limit=1)
        return s, request.env['survey.user_input'].sudo()

    answer_sudo = request.env['survey.user_input'].sudo().search([
        ('survey_id.access_token', '=', survey_token),
        ('access_token', '=', answer_token)
    ], limit=1)
    survey_sudo = answer_sudo.survey_id
    return survey_sudo, answer_sudo
```

flch-odoo replied:
Somehow `survey_id.access_token` didn't come to me at that time...
Please do have a look at the rest :)

std-odoo reviewed (commented):
@flch-odoo Hi :)

One comment while passing

Cheers

lm4649 commented on addons/survey/tests/test_certification_flow.py:
```diff
@@ -107,8 +112,12 @@ def test_flow_certification(self):
-            self._answer_question(q05, [q05.suggested_answer_ids.ids[0], q05.suggested_answer_ids.ids[1], q05.suggested_answer_ids.ids[3]], answer_token, csrf_token)
+            self._answer_question(q04, q04.suggested_answer_ids.ids[0], answer_token, csrf_token,
+                                  submit_query_count=47, access_page_query_count=30)
+            # Not the last to not add completion flow queries to answer validation
```
Could you clarify this comment above question 5? Does it mean that you shouldn't test query counts on the last question because of the additional completion flow? If so, why not (adapting expected number of queries)? 
Also, with a view to improving the experience of answering surveys, is it possible to optimize this additional completion flow?

lm4649 commented on addons/survey/models/survey_user_input.py:
```diff
@@ -778,23 +778,37 @@ def _check_answer_type_skipped(self):
     @api.model_create_multi
     def create(self, vals_list):
-        for vals in vals_list:
+        # Setup prefetching
```
typo in the commit title.

lm4649 reviewed (commented):
Hello @flch-odoo :wave: 

Great PR that deserves to be pushed forward in the pipe :+1: 

Why targeting version 17 ?

flch-odoo reviewed (commented):
Hey @lm4649 ,

Thanks for the comments. 
I chose 17.0 because many users are currently using this version, so if such improvements can be delivered to them, it's better. We try doing this when possible.
If it ends up unacceptable, I'll retarget easily.

Note: I could simplify a few lines using `nullcontext` and updated query counts in another file while I was here.

lm4649 commented on addons/survey/tests/test_certification_flow.py:
```diff
@@ -113,11 +113,11 @@ def test_flow_certification(self):
+                                  submit_query_count=32, access_page_query_count=23)
             self._answer_question(q06, q06.suggested_answer_ids.ids[0], answer_token, csrf_token,
-                                  submit_query_count=99, access_page_query_count=29)
+                                  submit_query_count=93, access_page_query_count=23)
```
note: submit_query_count: roughly a 7% performance gain

lm4649 reviewed (commented):
Hi @flch-odoo :wave: 

Thanks for answering my questions. I have added a few  minor/non-blocking questions / remarks.

17.0: I would have thought we should not modify the code of the prior stable version(s) unless something needs to be fixed.

Also I take this opportunity to pass on the recommendation to use an English spell checker when writing commits.

LGTM :+1:

awa-odoo reviewed (commented):
Hello @flch-odoo 

Neat little improvements!
Left a few comments & remarks.

Cheers!

awa-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -861,7 +878,11 @@ def _get_answer_score_values(self, vals, compute_speed_score=True):
                 suggested_answer_id = vals.get('suggested_answer_id')
                 if suggested_answer_id:
-                    question_answer = self.env['survey.question.answer'].browse(int(suggested_answer_id))
+                    question_answer = (
+                        suggested_answer_id
+                        if isinstance(suggested_answer_id, self.env['survey.question.answer'].__class__)
+                        else self.env['survey.question.answer'].browse(int(suggested_answer_id))
+                    )
```
I had missed this commit :upside_down_face: 

To simplify, I would make "_get_answer_score_values" take browserecords for "question", "user_input" and "suggested_answer".

And then in the create you would do:

```py
all_question_ids = [vals.get['question_id'] if vals.get('question_id') for vals in vals_list]
SurveyQuestion = self.env['survey_question'].with_prefetch(all_question_ids)
...
_get_answer_score_values({'question': SurveyQuestion.browse(question_id)})
```
(Hopefully clear, I skipped a few steps).

That way we leverage the power of "with_prefetch" and avoid instance checks that are rather dirty, not a fan of having a method taking either an ID or a browserecord for a param called "_id".

That is of course another approach that requires to change things in master rather than in 17.0.

Bertrand2 commented on addons/survey/controllers/main.py:
```diff
@@ -29,18 +29,24 @@ def _fetch_from_access_token(self, survey_token, answer_token):
+
+        return SurveySudo.with_context(active_test=False).search([('access_token', '=', survey_token)]), UserInputSudo
+
+    def _check_validity(self, survey_sudo, answer_sudo, answer_token, ensure_token=True, check_partner=True):
```
I'm not a big fan of changing the method signature in stable... Couldn't it be done using a context key? :thinking:

flch-odoo replied:
The target was changed to master ;-)

Bertrand2 replied:
Indeed! My bad, thanks :)

flch-odoo reviewed (commented):
Thank you @awa-odoo !

Of course, switching to master allows to simplify a few things.

(I'm currently working on a parallel branch to update query counts in each commit)

flch-odoo commented on addons/survey/tests/test_certification_flow.py:
```diff
@@ -107,8 +112,12 @@ def test_flow_certification(self):
-            self._answer_question(q04, q04.suggested_answer_ids.ids[0], answer_token, csrf_token)
-            self._answer_question(q05, [q05.suggested_answer_ids.ids[0], q05.suggested_answer_ids.ids[1], q05.suggested_answer_ids.ids[3]], answer_token, csrf_token)
+            self._answer_question(q04, q04.suggested_answer_ids.ids[0], answer_token, csrf_token,
+                                  submit_query_count=49, access_page_query_count=29)
```
Ah forgot this one. 
As it's not always a question (after final answer), it could be `access_next_page_query_count`. 
I'll change it at the next iteration if there is one :)

awa-odoo reviewed (commented):
Hello @flch-odoo 

Re-read this one, looking good but I still have some questions on the last commit.

Thanks for the update & cheers!

flch-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -785,23 +785,36 @@ def _check_answer_type_skipped(self):
+                score_vals = self._get_answer_score_values(
+                    vals
+                    | {'question': question, 'user_input': user_input}
+                    | ({'suggested_answer': suggested_answer} if suggested_answer else {})
```
We still need `vals` for all the value_X fields, but it's true we can be more explicit for the new values we use.

flch-odoo reviewed (commented):
Thank you for this new pass, @awa-odoo !

(last push was only documentation and moving `if answer_type == 'suggestion'` and `if suggested_answer` on the same line).

NB, for security (@odoo/rd-security :pray:) , the flags are related to these lines being moved. Security improvements are being introduced in #176890.

awa-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -786,24 +786,35 @@ def _check_answer_type_skipped(self):
+        suggested_answers = self.env['survey.question.answer'].browse(vals.get('suggested_answer_id', False) for vals in vals_list)
+        questions = self.env['survey.question'].browse(vals.get('question_id', False) for vals in vals_list)
+        user_inputs = self.env['survey.user_input'].browse(vals.get('user_input_id', False) for vals in vals_list)
+        if any(u.survey_id.session_speed_rating and u.is_session_answer for u in user_inputs):
+            existing_user_inputs = self.env['survey.user_input'].search([('id', 'in', user_inputs.ids)])
+            user_inputs = [user_input if user_input in existing_user_inputs else False for user_input in user_inputs]
```
Is this part really worth it VS the added complexity?

From what I understand, we prefetch user inputs, then we fetch all of their surveys (seems a bit counter-intuitive when trying to improve performances, but I'm guessing "_get_answer_score_values" does that further down the line), then check some conditions to do some extra processing.

That extra processing searches for user inputs, then changes the browserecord to a list based on existence comparison?

Which means you are zipping a mix of recordset and lists here below?

What are we gaining here exactly?

flch-odoo replied:
To be complete as it is still part of the new version: this was about checking once for the existence of all referenced user_inputs in flows that should count a speed adjusted score (instead of in a loop).

awa-odoo commented on addons/survey/models/survey_user_input.py:
```diff
@@ -785,23 +785,36 @@ def _check_answer_type_skipped(self):
     def create(self, vals_list):
-        for vals in vals_list:
+        # Setup prefetching
+        suggested_answers = self.env['survey.question.answer'].browse(vals.get('suggested_answer_id', False) for vals in vals_list)
+        questions = self.env['survey.question'].browse(vals.get('question_id', False) for vals in vals_list)
+        user_inputs = self.env['survey.user_input'].browse(vals.get('user_input_id', False) for vals in vals_list)
```
Ok, I see what you mean.

For me there is a big difference between using "for record in self", knowing the optimization that it implies, and building a recordset based on IDs, mixed with False values, then zipping multiple recordsets together.

Even if it uses the same mechanism, still think with_prefetch is more obvious (and would not require zipping).

But I will leave it to you to decide, probably nitpicking.

awa-odoo reviewed (commented):
Hello @flch-odoo 

Should be almost there :ping_pong: 

Thanks & cheers!

flch-odoo commented on addons/survey/controllers/main.py:
```diff
@@ -29,18 +29,24 @@ def _fetch_from_access_token(self, survey_token, answer_token):
+                    [('active', 'in', (True, False))]  # keeping active test for UserInput
+                ]))],
+                [('access_token', '=', answer_token)],
+            ]), limit=1)
```
If it doesn't cost much, it's:
1) just applying good practice, defaulting to the safest option
2) "Explicit is better than implicit" ("&")

flch-odoo commented:
Thank you for the review @awa-odoo . I've updated according to our discussion following the comments and had our beloved PO check that all went smoothly.

awa-odoo commented:
@robodoo r+ rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.

robodoo commented:
@flch-odoo @awa-odoo linked pull request(s) odoo/enterprise#73245 not ready. Linked PRs are not staged until all of them are ready.

flch-odoo commented:
Hello @awa-odoo , staging had failed on ENT, I just rebased and updated an initial query count, all is now green. Would you please repeat your r+ here and on [ENT](https://github.com/odoo/enterprise/pull/73245)?

awa-odoo commented:
Oula, this one is still alive, hum :eyes:

awa-odoo commented:
@robodoo r+
